unittest/server/src/ivi_controller_uinttests.cpp

```diff
@@ -0,0 +1,3475 @@
+    ASSERT_EQ(lp_iviScreen->id_screen, mp_screenInfo->screen_id);
+
+    free(lp_iviScreen);
+}
+
+/** ================================================================================================
+ * @test_id             bind_ivi_controller_nullResource
+ * @brief               Test case of bind_ivi_controller() where wl_resource_create() fails, return null pointer
+ * @test_procedure Steps:
+ *                      -# Calling the bind_ivi_controller()
+ *                      -# Verification point:
+ *                         +# wl_resource_create() must be called once time
+ *                         +# wl_resource_set_implementation() not be called
+ */
+TEST_F(ControllerTests, bind_ivi_controller_nullResource)
+{
+    // update version >= 2
+    bind_ivi_controller(mp_fakeClient, mp_iviShell, 2, 1);
+
+    ASSERT_EQ(wl_resource_create_fake.call_count, 1);
+    ASSERT_EQ(wl_resource_set_implementation_fake.call_count, 0);
+}
+
+/** ================================================================================================
+ * @test_id             bind_ivi_controller_success
+ * @brief               Test case of bind_ivi_controller() where wl_resource_create() success, return an object
+ * @test_procedure Steps:
+ *                      -# Mocking the wl_resource_create() does return an object
+ *                      -# Calling the bind_ivi_controller()
+ *                      -# Verification point:
+ *                         +# wl_resource_create() must be called once time
+ *                         +# wl_resource_set_implementation() must be called once time
+ *                         +# wl_list_insert() must be called once time
+ *                         +# wl_list_init() must be called 2 times
+ *                         +# get_id_of_surface() must be called {MAX_NUMBER} times
+ *                         +# get_id_of_layer() must be called {MAX_NUMBER} times
+ *                         +# wl_resource_post_event() must be called {MAX_NUMBER*2} times
+ *                         +# The result output should same with prepare data
+ *                         +# Free resources are allocated when running the test
+ */
+TEST_F(ControllerTests, bind_ivi_controller_success)
+{
+    SET_RETURN_SEQ(wl_resource_create, &mp_wlResourceDefault, 1);
+
+    // update version >= 2
+    bind_ivi_controller(mp_fakeClient, mp_iviShell, 2, 1);
+
+    ASSERT_EQ(wl_resource_create_fake.call_count, 1);
+    ASSERT_EQ(wl_resource_set_implementation_fake.call_count, 1);
+    ASSERT_EQ(wl_list_insert_fake.call_count, 1);
+    ASSERT_EQ(wl_list_init_fake.call_count, 2);
+    ASSERT_EQ(get_id_of_surface_fake.call_count, MAX_NUMBER);
+    ASSERT_EQ(get_id_of_layer_fake.call_count, MAX_NUMBER);
+    ASSERT_EQ(wl_resource_post_event_fake.call_count, MAX_NUMBER*2);
+    struct ivicontroller *lp_iviController = (struct ivicontroller*)(uintptr_t(wl_list_init_fake.arg0_history[0]) - offsetof(struct ivicontroller, surface_notifications));
```

Comment:
good to get lp_iviController via wl_list_insert, we should care the object was inserted to list of controllers